Add checklists + restructure instructions (#132)
knative/release: PROCEDURES.md

## Build health check
tl;dr: check that all builds on `main` are passing.

Check if the repository is in a good shape and the builds pass consistently. **This is required** because the [Prow job](#release-job) that builds the release artifacts will execute all the various tests (build, unit, e2e) and, if something goes wrong, the release process will probably need to restart from the beginning.

For any problems in a specific repo, get in touch with the relevant WG leads to fix them.

## Dependency check
tl;dr: check that all the dependencies for the repo are up-to-date and aligned with the relase version.

### Exceptions
Repos that don't have dependencies naturally don't need a dependency check and this step can be skipped for those. Currently, `knative.dev/hack` is the only `knative.dev` repo that does not have any dependencies.

### Aligning dependencies
Each repo needs to be successfully updated to use the latest version of all shared dependencies **before** its release branch is cut.

> In order to align the `knative.dev` dependencies, `knative-sandbox/knobots` automation will run PRs like this [Upgrade to latest dependencies PR](https://github.com/knative/eventing/pull/4713) for each repo, executing the command `./hack/update-deps.sh --upgrade --release 0.20` and committing all the content. Note: `buoy check`, which is invoked in the script, will fail if the dependencies are not yet ready.

Reviewer: ```suggestion
In order to align the `knative.dev` dependencies, `knative-sandbox/knobots` automation will run "Upgrade to latest dependencies PRs ([example](https://github.com/knative/eventing/pull/4713) for each repo, executing the command `./hack/update-deps.sh --upgrade --release 0.20` and committing all the content. Note: `buoy check`, which is invoked in the script, will fail if the dependencies are not yet ready.
```

> - If there is no "Upgrade to latest dependencies" PR open, manually trigger the generation of these PRs starting the [Knobots Auto Update Deps](https://github.com/knative-sandbox/knobots/actions/workflows/auto-update-deps.yaml) and wait for the PR to pop in the repo you need. Note that in the automation run you have to change the field `If true, send update PRs even for deps changes that don't change vendor. nUse this only for releases.` to **true**, because in some cases there are no code changes in the vendor.

Author: Yes, thank you for that.

> - If there is no "Upgrade to latest dependencies" PR open, the update PR might already have been merged. If this is not the case, manually trigger the generation of this PR starting the [Knobots Auto Update Deps](https://github.com/knative-sandbox/knobots/actions/workflows/auto-update-deps.yaml) and wait for the PR to pop in the repo you need. Note that in the automation run you have to change the field `If true, send update PRs even for deps changes that don't change vendor. nUse this only for releases.` to **true**, because in some cases there are no code changes in the vendor.

Reviewer: ```suggestion
- If there is no "Upgrade to latest dependencies" PR open, the update PR might already have been merged. If this is not the case, manually trigger the generation of this PR starting the [Knobots Auto Update Deps](https://github.com/knative-sandbox/knobots/actions/workflows/auto-update-deps.yaml) and wait for the PR to pop in the repo you need. Note that in the automation run you have to change the field `If true, send update PRs even for deps changes that don't change vendor. Use this only for releases.` to **true**, because in some cases there are no code changes in the vendor.
```

- [Update releasability defaults for 1.5 by gab-satchi · Pull Request #202 · knative-sandbox/knobots](https://github.com/knative-sandbox/knobots/pull/202?w=1)

The changes in the source workflow will be automatically propagated to the rest of Knative repos in the next round of
workflow syncs, and it can also be triggered manually when necessary.

### Verifying releasability
An automatic repo releasability test is executed periodically and the results are posted on the corresponding Slack release channel. If the dependencies are properly aligned, the releasability test will pass.

A test can be re-run by manually running the [Releasability workflow](https://github.com/knative/serving/actions?query=workflow%3AReleasability).

If the releasability test reports a NO-GO on a repo where it was previously passing, probably a new PR merge introduced a dependency misalignment. Whatever the reason, if the releasability test is failing, it is necessary to start with the first step in the [Releasing a repository](#releasing-a-repository) process to get this test to a passing state.

## Cut a branch

> tl;dr: cut a release branch from `main`.
> ### Exceptions

Reviewer: ```suggestion
tl;dr: cut a release branch from `main`.

### Exceptions
```

For some repositories some extra manual validation and updates need to be performed before the release branch is cut:

- `knative/client`
  - (optional) Check that [CHANGELOG.adoc](https://github.com/knative/client/blob/main/CHANGELOG.adoc) contains a section about the release, i.e. the top-level "(Unreleased)" section should be changed to point to the upcoming release version and number. It's not critical if the changelog is aligned after the release in retrospective.
  - If the validation fails, the fix should be trivial and could be either performed by the release leads or the client team.

- `knative-sandbox/kn-plugin-quickstart`

>   - Update the version numbers of Serving / Kourier / Eventing in [pkg/install/install.go](https://github.com/knative-sandbox/kn-plugin-quickstart/blob/main/pkg/install/install.go#L25-L27) so that the plugin will use the just-released versions.

Reviewer: ```suggestion
  - Update the version numbers of Serving / Kourier / Eventing in [pkg/install/install.go](https://github.com/knative-sandbox/kn-plugin-quickstart/blob/4e57ed10eaac8882b0a81b2861ddaee37bfcd6f9/pkg/install/install.go#L25-L27) so that the plugin will use the just-released versions.
```

Probably better to pin to a specific commit, as the line references might change if updates are made (this one's on me... I added this line a while back and really should've used an explicit reference :facepalm: )

### Cutting a branch
Cutting a `release-x.y` branch can be done by using the GitHub UI:

1. Click on the branch selection box at the top level page of the repository.

  ![Click the branch selection box](images/github-branch.png)

> 2. Search for the correct `release-x.y` branch name for the release.

Reviewer: ```suggestion
1. Search for the correct `release-x.y` branch name for the release.
```

2. ask the relevant WG leads if they have any add or edit for the updated repo notes in the [HackMD document](https://hackmd.io/cJwvzJ4eRVeqqiXzOPtxsA)
3. use the GitHub UI to update the repo release tag with the final notes

## Homebrew updates
### homebrew-client
After the client release, the [Homebrew tap](https://github.com/knative/homebrew-client) needs to be updated with the new release:

- Copy `kn.rb` to the `kn@${PREV_RELEASE}.rb` with `$PREV_RELEASE` to be replace with the latest release (e.g. `0.19`).
- In `kn@${PREV_RELEASE}.rb` replace `class Kn` with `class KnAT${PREV_RELEASE_DIGITS}`, e.g `class KnAT019` for an previous release `0.19`.
- In `kn.rb`:
  - Replace the old version number in `v` with the released version (e.g. `v = "v0.20.0"`)
  - Replace the `sha256` checksums with the values from the [client release page](https://github.com/knative/client/releases). The checksums have been released too (e.g. [checksums.txt](https://github.com/knative/client/releases/download/v0.22.0/checksums.txt))

✅ Open a PR and merge the changes. Prow is not enabled for the homebrew repo, so the merge needs to be performed manually.

> ### homebrew-kn-plugins

Reviewer: I just added a script to automate the version bump for the homebrew plugin formulas: https://github.com/knative-sandbox/homebrew-kn-plugins/blob/main/hack/update-versions.sh  Going to add one for the client formula also... hopefully makes this process a little easier
